add python3.9 testing and related llfuse changes (backport to 1.1-maint)

ernestoarbitrio opened the pull request:
Includes PRs: https://github.com/borgbackup/borg/pull/5373 https://github.com/borgbackup/borg/pull/5379 
It also includes this specific commit https://github.com/borgbackup/borg/commit/7dc161067498703dca27b6e2437548a9cc4d981b

ThomasWaldmann commented on setup.py:
```diff
@@ -52,16 +52,19 @@
 
 if sys.platform.startswith('freebsd'):
     # llfuse was frequently broken / did not build on freebsd
     # llfuse 0.41.1, 1.1 are ok
     extras_require['fuse'] = ['llfuse <2.0, !=0.42.*, !=0.43, !=1.0', ]
 
-if my_python >= (3, 7):
-    extras_require['fuse'][0] += ', >=1.3.4'
+# if my_python >= (3, 7):
+#     extras_require['fuse'][0] += ', >=1.3.4'
```
good find! yeah, this did not match the code above any more.
guess the freebsd change right above also needs removal.

2 possibilities now:
- open a ticket about it and later fix master and 1.1-maint in a separate PR.
- have a bit of scope creep (this fix is not a backport, but a new change!) and remove (not: comment) the code here.

ernestoarbitrio replied:
mmm i think that master is ok ... your changes reflect what i changed in this branch

ThomasWaldmann replied:
right, master is ok, no such code there. so just remove it here.

ThomasWaldmann commented on tox.ini:
```diff
@@ -2,7 +2,7 @@
 
 [tox]
-envlist = py{34,35,36,37,38},flake8
+envlist = py{35,36,37,38,39},flake8
```
don't drop py34 here, we still announce support for it in pypi package metadata.

(guess I will drop it **after** 1.1.14 release though)

ThomasWaldmann reviewed (commented):
thanks for the backporting!

these are a bit more commits than expected though, i can help you cleaning up later if you like.

keep commits small and clean ("1 idea per commit") from now on, that will ease cleaning up later.

ernestoarbitrio commented:
@ThomasWaldmann i cherry-picked all my commits from the previous PR and your commits mentioned in https://github.com/borgbackup/borg/pull/5373. When you say `cleaning up` you mean create a new PR and pick only some of those commits?

ernestoarbitrio commented on .travis.yml:
```diff
@@ -27,14 +27,34 @@ matrix:
           os: linux
           dist: xenial
           env: TOXENV=py37
+        - python: "3.7-dev"
+          os: linux
+          dist: xenial
+          env: TOXENV=py37
```
?? you mean that in this branch we should remove the `*-dev`?

ThomasWaldmann replied:
guess they were not there in 1.1-maint before you added them?

ThomasWaldmann replied:
so, "yes".

ernestoarbitrio replied:
yes u right getting the commits i added them ... i'lll remove

ThomasWaldmann commented:
I meant the fix commits, e.g. "fix conflicts".

With cleaning up I mean to squash together A and B, if B is a fixup commit for A (e.g. fixing a merge conflict).
One can (very carefully) use `git rebase -i 1.1-maint` for that.

ernestoarbitrio commented:
Ok i think i addressed your comments, if you can take a further look it would be great. Even what i got from the dev guide about cherry-picking stuff confused me ... i'd be more focused when getting old commits in backports branchs. Any suggestion is appreciated :D

ThomasWaldmann commented on .travis/install.sh:
```diff
@@ -27,8 +26,12 @@ then
-        pyenv install 3.5.3  # Minimum version for OpenSSL 1.1.x
-        pyenv global 3.5.3
+        pyenv install 3.5.0
+        pyenv global 3.5.0
```
see comment, that must stay at 3.5.3.

ThomasWaldmann commented on .travis/install.sh:
```diff
@@ -27,8 +26,12 @@ then
     if [ "${TOXENV}" = "py35" ]
     then
-        pyenv install 3.5.3  # Minimum version for OpenSSL 1.1.x
+        pyenv install 3.5.3
```
please keep the comment as a reminder not to use anything less that 3.5.3 here.

e.g. 3.5.0 just won't compile with openssl 1.1.

ThomasWaldmann reviewed (commented):
btw, when doing changes, do you read the current and overall diff yourself?

like: `git diff` and `git diff 1.1-maint` ?

ThomasWaldmann commented:
Tell me when you're finished, then I'll just squash together all commits and merge.

First I thought there was something different in the "resolve conflicts" commits, but guess like it is now, it is too much effort to clean up using git rebase -i (and have a result of a few clean commits). So we'll just have 1 commit with everything and make sure that one is ok.

ThomasWaldmann commented:
@ernestoarbitrio you asked for typical backports that could act as an example.

Well, there is not much to see. If one is lucky, the backport will look exactly the same as the original PR, just apply the same change to the "older" branch (1.1-maint). Similar is a forward port applying a change to the newer branch (master).

In case cherry-picking complains about a conflict, things get more interesting (like you likely had it here).
The cherry-pick operation then does not complete, but just tells you to fix the conflict manually (and also what to do afterwards).

The idea then is not to just take the new code, but to apply the same change (same idea) the new code got in the changeset one is backporting to the old code. Not more and not less.
After fixing manually, one completes the cherry-pick as git tells (and it'll automatically take the same commit comment) and it will result in 1 commit.

As always, one should try not to add scope creep. The scope is just doing the same change / implement the same idea as in the orginal changeset. Sometimes this requires more changes to not break stuff.

But one tries to not to do any unrelated change (e.g. if one spots a bug somewhere else). This is also known as a "clean changeset".

ThomasWaldmann commented on .travis/install.sh:
```diff
@@ -27,8 +26,12 @@ then
-        pyenv install 3.5.3  # Minimum version for OpenSSL 1.1.x
+        pyenv install 3.5.3
         pyenv global 3.5.3
+    elif [ "${TOXENV}" = "py36" ]
+    then
+        pyenv install 3.6.0
+        pyenv global 3.6.0
```
is this used?

ernestoarbitrio replied:
not actually

ThomasWaldmann reviewed (commented):
i fixed a minor nitpick myself and will squash/merge this after travis-ci completed (assuming there are no new problems).

thanks for working on it!

ernestoarbitrio commented:
> i fixed a minor nitpick myself and will squash/merge this after travis-ci completed (assuming there are no new problems).
> 
> thanks for working on it!

Nice! Thanks :D
